raster/src/main/scala/geotrellis/raster/io/geotiff/GeoTiffMultibandTile.scala

```diff
@@ -166,74 +168,36 @@ abstract class GeoTiffMultibandTile(
+          getSegments(0 until segmentCount).foreach { case (segmentIndex, segment) =>
             val (cols, rows) =
-              if(segmentLayout.isTiled) { (segmentLayout.tileLayout.tileCols, segmentLayout.tileLayout.tileRows) }
-              else { segmentLayout.getSegmentDimensions(segmentIndex) }
-
-            val paddedCols = {
-              val bytesWidth = (cols + 7) / 8
-              bytesWidth * 8
-            }
-
-            val resultByteCount =
-              (paddedCols / 8) * rows
-
-            val bandSegment = Array.ofDim[Byte](resultByteCount)
-            val segment = getSegment(segmentIndex)
-
-            cfor(bandIndex)(_ < segment.size, _ + bandCount) { i =>
-              val j = i / bandCount
-
-              val col = (j % cols)
-              val row = (j / cols)
-              val i2 = (row * paddedCols) + col
-
-              BitArrayTile.update(bandSegment, i2, segment.getInt(i))
-            }
-
-            // Inverse the byte, to account for endian mismatching.
-            cfor(0)(_ < bandSegment.size, _ + 1) { i =>
-              bandSegment(i) = invertByte(bandSegment(i))
-            }
-
-            compressedBandBytes(segmentIndex) = compressor.compress(bandSegment, segmentIndex)
+              if (segmentLayout.isTiled) (segmentLayout.tileLayout.tileCols, segmentLayout.tileLayout.tileRows)
+              else segmentLayout.getSegmentDimensions(segmentIndex)
+            val bytes = GeoTiffSegment.deinterleaveBitSegment(segment, cols, rows, bandCount)(bandIndex)
+            compressedBandBytes(segmentIndex) = compressor.compress(bytes, segmentIndex)
           }
 
           GeoTiffTile(new ArraySegmentBytes(compressedBandBytes), compressor.createDecompressor(), segmentLayout, compression, cellType, Some(bandType))
         case _ =>
           val compressedBandBytes = Array.ofDim[Array[Byte]](segmentCount)
           val compressor = compression.createCompressor(segmentCount)
           val bytesPerSample = bandType.bytesPerSample
-          val bytesPerCell = bytesPerSample * bandCount
-
-          cfor(0)(_ < segmentCount, _ + 1) { segmentIndex =>
-            val segment = getSegment(segmentIndex).bytes
-            val segmentSize = segment.size
-            val bandSegmentCount = segmentSize / bandCount
-            val bandSegment = Array.ofDim[Byte](bandSegmentCount)
-
-            var j = 0
-            cfor(bandIndex * bytesPerSample)(_ < segmentSize, _ + bytesPerCell) { i =>
-              var b = 0
-              while(b < bytesPerSample) { bandSegment(j + b) = segment(i + b) ; b += 1 }
-              j += bytesPerSample
-            }
 
-            compressedBandBytes(segmentIndex) = compressor.compress(bandSegment, segmentIndex)
+          getSegments(0 until segmentCount).foreach { case (segmentIndex, geoTiffSegment) =>
+            val bytes = GeoTiffSegment.deinterleave(geoTiffSegment.bytes, bandCount, bytesPerSample)(bandIndex)
```

Comment:
This is actually not very efficient because we know we only want one band, but these functions will de-interlace all present bands, allocating all that memory for nothing.